File: Day02/main.py
def read_reports() -> list[list[int]]:
    with open("input.txt") as file:
        lines = file.read().rstrip().split('\n')
    def convert_line(line):
        return list(map(int, line.split()))
    reports = list(map(convert_line, lines))
    return reports

def is_safe_increasing(report: list[int]) -> bool:
    if len(report) <= 1:
        return True
    for i in range(len(report) - 1):
        if not (report[i] + 1 <= report[i+1] <= report[i] + 3):
            return False
    return True

def part1():
    reports = read_reports()
    def is_safe(report: list[int]) -> bool:
        return is_safe_increasing(report) or is_safe_increasing(report[::-1])
    num_safe_reports = len([report for report in reports if is_safe(report)])
    print(num_safe_reports)

def part2():
    reports = read_reports()
    def tolerant_is_safe_increasing(report: list[int]) -> bool:
        if is_safe_increasing(report):
            return True
        else:
            for i in range(len(report)):
                if is_safe_increasing(report[:i] + report[i+1:]):
                    return True
            return False
    def tolerant_is_safe(report: list[int]) -> bool:
        return tolerant_is_safe_increasing(report) or tolerant_is_safe_increasing(report[::-1])
    num_safe_reports = len([report for report in reports if tolerant_is_safe(report)])
    print(num_safe_reports)    
    
if __name__ == '__main__':
    part1()
    part2()

File: Day02/main.rs
use std::fs;

fn read_input_as_rows() -> Vec<Vec<i32>> {
    fs::read_to_string("input.txt")
        .unwrap()
        .lines()
        .map(|line| {
            line.split(" ")
                .map(|word| word.parse::<i32>().unwrap())
                .collect()
        })
        .collect()
}

fn is_safe_inc<'a, I>(mut report: I) -> bool
where
    I: Iterator<Item = &'a i32>,
{
    let x = report.next();
    if x.is_none() {
        return true;
    }
    let mut x = x.unwrap();
    for y in report {
        if !(1 <= y - x && y - x <= 3) {
            return false;
        }
        x = y;
    }
    return true;
}

fn is_safe<'a, I>(report: I) -> bool
where
    I: DoubleEndedIterator<Item = &'a i32> + Clone,
{
    is_safe_inc(report.clone()) || is_safe_inc(report.rev())
}

fn part1() {
    let reports = read_input_as_rows();
    let num_safe_reports = reports
        .iter()
        .filter(|&report| is_safe(report.iter()))
        .count();
    println!("{}", num_safe_reports);
}

fn is_safe_excluding_one(report: &Vec<i32>) -> bool {
    for i in 0..report.len() {
        let report_excluding_i = report
            .iter()
            .enumerate()
            .filter(|(j, _)| i != *j)
            .map(|(_, x)| x);
        if is_safe(report_excluding_i) {
            return true;
        }
    }
    return false;
}

fn part2() {
    let reports = read_input_as_rows();
    let num_safe_reports = reports
        .iter()
        .filter(|&report| is_safe(report.iter()) || is_safe_excluding_one(report))
        .count();
    println!("{}", num_safe_reports);
}

fn main() {
    part1();
    part2();
}
